packages/compiler/core/util.ts

```diff
@@ -185,6 +200,28 @@ export async function findProjectRoot(
   }
 }
 
+/**
+ * The mutable equivalent of a type.
+ */
+//prettier-ignore
+export type Mutable<T> =
+  T extends SymbolTable ? T & { set(key: string, value: Sym): void } :
+  T extends ReadonlyMap<infer K, infer V> ? Map<K, V> :
+  T extends ReadonlySet<infer T> ? Set<T> :
+  T extends readonly (infer V)[] ? V[] :
+  // brand to force explicit conversion.
+  { -readonly [P in keyof T]: T[P] } & { __writableBrand: never };
+
+/**
+ * Casts away readonly typing.
+ *
+ * Use it like this when it is safe to override readonly typing:
+ *   mutate(item).prop = value;
+ */
+export function mutate<T>(value: T): Mutable<T> {
```

Comment:
This only cheat the typescript system right? If it has been frozen there is no way back?

Reply:
Correct. 

We don't actually freeze anything at this time. I did some experiments with freezing to find issues, but nothing I intend to check-in. The reuse will rely on not cheating the TypeScript typing, which I think is OK.

Reply:
In theory, we could freeze after parse-and-bind and then you would get crashes if you tried to use mutate() after the freezing. That would be nice, but I'm not sure if the expense is worth it and also it breaks things like lazily setting flags. I think we would have to use Object.defineProperty with writable false on everything but flags instead of Object.freeze.